File: backend/app/schemas/audit.py
"""Pydantic schemas for audit log API requests and responses."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class AuditLogResponse(BaseModel):
    """Single audit log entry returned by the API."""

    id: uuid.UUID
    timestamp: datetime
    action: str
    category: str
    actor_email: str | None = None
    actor_role: str | None = None
    resource_type: str | None = None
    resource_id: str | None = None
    resource_name: str | None = None
    details: dict | None = None
    ip_address: str | None = None
    status: str
    error_message: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AuditLogListResponse(BaseModel):
    """Paginated list of audit log entries."""

    logs: list[AuditLogResponse]
    total: int
    page: int
    page_size: int


class AuditQueryParams(BaseModel):
    """Query parameters for filtering and paginating audit logs."""

    page: int = Field(1, ge=1, description="Page number (1-indexed)")
    page_size: int = Field(50, ge=1, le=500, description="Results per page")
    action: str | None = Field(None, description="Filter by action (e.g. server.created)")
    category: str | None = Field(None, description="Filter by category (e.g. server, auth)")
    actor_email: str | None = Field(None, description="Filter by actor email")
    resource_id: str | None = Field(None, description="Filter by resource UUID")
    start_date: datetime | None = Field(None, description="Include logs on or after this timestamp")
    end_date: datetime | None = Field(None, description="Include logs on or before this timestamp")
    status: str | None = Field(None, description="Filter by outcome status (success, failure, error)")


class AuditDailyCounts(BaseModel):
    """Audit events aggregated by day."""

    date: str
    count: int


class AuditActionCount(BaseModel):
    """Count of events for a specific action type."""

    action: str
    count: int


class AuditActorCount(BaseModel):
    """Count of events for a specific actor."""

    actor_email: str
    count: int


class AuditSummaryResponse(BaseModel):
    """Aggregate statistics for the audit dashboard."""

    daily_counts: list[AuditDailyCounts]
    top_actions: list[AuditActionCount]
    top_actors: list[AuditActorCount]
    total_events: int

File: frontend/src/lib/api/audit.ts
/**
 * Audit log API types, endpoint functions, and React Query hooks.
 *
 * Backend routes:
 *   GET /api/v1/audit          — List audit logs (paginated, filterable)
 *   GET /api/v1/audit/{id}     — Single audit log entry
 *   GET /api/v1/audit/actions  — Distinct action types
 *   GET /api/v1/audit/summary  — Aggregate stats (events/day, top actions, top actors)
 */

import { useQuery } from '@tanstack/react-query';
import apiClient from './client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** A single audit log entry. */
export interface AuditLog {
  id: string;
  timestamp: string;
  action: string;
  category: string;
  actor_email: string | null;
  actor_role: string | null;
  resource_type: string | null;
  resource_id: string | null;
  resource_name: string | null;
  details: Record<string, unknown> | null;
  ip_address: string | null;
  status: string;
  error_message: string | null;
  created_at: string;
}

/** Paginated audit log list response. */
export interface AuditLogListResponse {
  logs: AuditLog[];
  total: number;
  page: number;
  page_size: number;
}

/** Aggregate audit stats returned by the summary endpoint. */
export interface AuditSummary {
  events_per_day: { date: string; count: number }[];
  top_actions: { action: string; count: number }[];
  top_actors: { email: string; count: number }[];
  total_events: number;
}

/** Query parameters accepted by the list endpoint. */
export interface AuditQueryParams {
  page?: number;
  page_size?: number;
  action?: string;
  category?: string;
  actor_email?: string;
  resource_id?: string;
  start_date?: string;
  end_date?: string;
  status?: string;
}

// ---------------------------------------------------------------------------
// Endpoint functions
// ---------------------------------------------------------------------------

/** Fetch a paginated, filtered list of audit log entries. */
export async function listAuditLogs(
  params?: AuditQueryParams,
): Promise<AuditLogListResponse> {
  const response = await apiClient.get<AuditLogListResponse>('/audit', {
    params,
  });
  return response.data;
}

/** Fetch a single audit log entry by ID. */
export async function getAuditLog(id: string): Promise<AuditLog> {
  const response = await apiClient.get<AuditLog>(`/audit/${id}`);
  return response.data;
}

/** Fetch the list of distinct action types recorded in the audit log. */
export async function getAuditActions(): Promise<string[]> {
  const response = await apiClient.get<string[]>('/audit/actions');
  return response.data;
}

/** Fetch aggregate audit summary (events/day, top actions, top actors). */
export async function getAuditSummary(): Promise<AuditSummary> {
  const response = await apiClient.get<AuditSummary>('/audit/summary');
  return response.data;
}

// ---------------------------------------------------------------------------
// Query key factory
// ---------------------------------------------------------------------------

export const auditQueryKeys = {
  all: ['audit'] as const,
  list: (params?: AuditQueryParams) => ['audit', 'list', params] as const,
  detail: (id: string) => ['audit', id] as const,
  actions: ['audit', 'actions'] as const,
  summary: ['audit', 'summary'] as const,
} as const;

// ---------------------------------------------------------------------------
// React Query hooks
// ---------------------------------------------------------------------------

/**
 * Fetch a paginated, filtered audit log list.
 *
 * Accepts an optional `refetchInterval` so callers can enable auto-refresh.
 */
export function useAuditLogs(
  params?: AuditQueryParams,
  options?: { refetchInterval?: number | false },
) {
  return useQuery<AuditLogListResponse>({
    queryKey: auditQueryKeys.list(params),
    queryFn: () => listAuditLogs(params),
    refetchInterval: options?.refetchInterval,
  });
}

/** Fetch a single audit log entry. Disabled when `id` is undefined. */
export function useAuditLog(id: string | undefined) {
  return useQuery<AuditLog>({
    queryKey: auditQueryKeys.detail(id!),
    queryFn: () => getAuditLog(id!),
    enabled: !!id,
  });
}

/** Fetch the distinct action types. Cached aggressively since these change rarely. */
export function useAuditActions() {
  return useQuery<string[]>({
    queryKey: auditQueryKeys.actions,
    queryFn: getAuditActions,
    staleTime: 5 * 60 * 1000, // 5 minutes
  });
}

/** Fetch the aggregate audit summary. */
export function useAuditSummary() {
  return useQuery<AuditSummary>({
    queryKey: auditQueryKeys.summary,
    queryFn: getAuditSummary,
  });
}
